force_bdss/data_sources/base_data_source_model.py

```diff
@@ -46,27 +50,59 @@ class BaseDataSourceModel(BaseModel):
     #: this and adapt the visual entries.
     changes_slots = Event()
 
-    def verify(self):
-        """ Verify the data source model.
+    def __init__(self, factory=None, *args, **kwargs):
+        super(BaseDataSourceModel, self).__init__(factory, *args, **kwargs)
+        # Perform extra checks on input_slot_info and output_slot_info
+        # to make sure they will be accepted by the BaseDataSource associated
+        # with factory
+        self._assign_slot_info()
 
-        The data source model must have:
-        - input and output slots match between instance and model
-        - all output slots named
-        - no errors in input or output slots
+    # -------------------
+    #      Listeners
+    # -------------------
+
+    @on_trait_change("+changes_slots")
```

Comment:
I would appreciate if you could tell me how this works (and the `ChangesSlotsModel` in `test_base_data_source_model.py` as well). Thank you!

Reply:
Sure, this `on_trait_change` handler listens to all traits on the `BaseDataSourceModel` class that have the `changes_slots=True` metadata and is called whenever one of them is changed.

Therefore a developer should make sure that any trait on a `BaseDataSourceModel` subclass that is expected to cause a change in the format of the return values of the data source `slots` method has the `changes_slots=True` metadata.

The `ChangesSlotsModel` is an example subclass, which should only experience a `_trigger_changes_slots` method call when the `b` trait is changed

Reply:
The `_trigger_changes_slots` method uses the fact that the `on_trait_change` decorator mutates the signature of its method to include objects corresponding to the updated trait: https://docs.enthought.com/traits/traits_user_manual/notification.html#dynamic-notification

    handler(object, name, old, new)